core/src/main/java/com/godaddy/commerce/core/jwt/CertBasedJwtCache.java

```diff
@@ -0,0 +1,117 @@
+package com.godaddy.commerce.core.jwt;
+
+import com.godaddy.commerce.core.errors.ProcessingException;
+import com.godaddy.commerce.core.Expirable;
+import com.godaddy.commerce.core.http.config.ConnectionProperty;
+import com.godaddy.commerce.core.http.config.EcommerceHost;
+import com.godaddy.commerce.core.http.HttpVerbWrapper;
+import io.vertx.ext.web.RoutingContext;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.HttpStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public enum CertBasedJwtCache {
+  INSTANCE;
+
+  private final static Logger LOGGER = LoggerFactory.getLogger(CertBasedJwtCache.class);
+  private final static Map<String, Expirable<JwtResponse>> CACHE = new HashMap<>();
+  private final static String SSO_TOKEN_HTTP_PATH = "/v1/secure/api/token";
+  private static String SSO_CONNECTION_PROPERTY_KEY = "SSO";
+
+  public static void setEnvType(String envType) {
+    // Use suffixed key for SSO in PCI/ICP.
+    SSO_CONNECTION_PROPERTY_KEY = envType == "PCI" || envType == "ICP"
+      ? String.format("SSO_%s", envType)
+      : "SSO";
+  }
+
+  private static boolean isValid(Expirable<JwtResponse> jwt) {
+    return jwt != null && !jwt.isExpired();
+  }
+
+  public String get(RoutingContext routingContext, String certId) {
+    if (StringUtils.isBlank(certId)) {
+      throw new IllegalArgumentException("Cert ID required.");
+    }
+
+    // Use double-checked locking to allow concurrent gets of fresh JWTs while also preventing
+    // redundant writes once it becomes stale.
+
+    Expirable<JwtResponse> jwt = CACHE.get(certId);
+
+    if (!isValid(jwt)) {
+      synchronized (this) {
+        jwt = CACHE.get(certId);
+
+        if (!isValid(jwt)) {
+          try {
+            final ConnectionProperty ssoConnectionProperty = EcommerceHost.INSTANCE
+                .connectionProperties().get(SSO_CONNECTION_PROPERTY_KEY);
+
+            // As an external service, SSO can only be called with `url`.
+            if (ssoConnectionProperty == null || StringUtils.isBlank(ssoConnectionProperty.url())) {
+              throw new RuntimeException(
+                  String.format(
+                      "%s connection property is missing URL.",
+                      SSO_CONNECTION_PROPERTY_KEY
+                  )
+              );
+            }
+
+            String ssoUrl = ssoConnectionProperty.url() + SSO_TOKEN_HTTP_PATH;
+
+            JwtResponse response = HttpVerbWrapper.instance()
+                .routingContext(routingContext)
+                .useSslContext(certId)
+                .contentType("application/x-www-form-urlencoded")
+                .post(JwtResponse.class, ssoUrl, "realm=cert")
+                .expectsCode(HttpStatus.SC_CREATED, "jwtTokenCreated")
+                .expectsResult("jwtTokenCreated")
+                .result();
+
+            if (response.message() != "Success") {
```

Comment:
> [<img alt="szhang1" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.secureserver.net/szhang1) **Authored by [szhang1](https://github.secureserver.net/szhang1)**
_<time datetime="2020-11-16T18:05:07Z" title="Monday, November 16th 2020, 7:05:07 pm +01:00">Nov 16, 2020</time>_

---


```suggestion
            if (!"Success".equals(response.message())) {
```